ocaml/xapi/xapi_pgpu.ml

```diff
@@ -367,17 +367,76 @@ let mxgpu_vf_setup ~__context =
 (* This must be run LOCALLY on the host that is about to start a VM that is
  * going to use a vgpu backed by an nvidia pgpu. *)
 let nvidia_vf_setup ~__context ~pf ~enable =
```

Comment:
Does this function need some locking? e.g. in message-forwarding
Although the actions performed by it are idempotent you could have 2 threads racing to enable the extra VFs when number of VFs is zero, which may not work well.

Reply:
I agree, especially now that multiple VM-starts can get here at the same time. I think that a local mutex in this function would suffice.